File: main/scala/mod/iceandshadow3/compat/HarvestMethod.java
package mod.iceandshadow3.compat;

import javax.annotation.Nullable;

public class HarvestMethod {
	private static java.util.HashMap<String, HarvestMethod> lookups;
	public static HarvestMethod
		PICKAXE = new HarvestMethod("pickaxe"),
		SHOVEL = new HarvestMethod("shovel"), //Let's not have a spade repeat.
		AXE = new HarvestMethod("axe"),
		BLADE = new HarvestMethod("sword"),
		SHEAR = new HarvestMethod("shear", null);
	public static HarvestMethod get(String name) {
		HarvestMethod retval = lookups.get(name);
		if(retval == null) retval = new HarvestMethod(name);
		return retval;
	}
	
	protected String name;
	private HarvestMethod(String string) {
		this(string, string);
	}
	private HarvestMethod(String name, @Nullable String classname) {
		this.name = name;
		if(classname != null) lookups.putIfAbsent(classname, this);
	}
	
	@Override
	public String toString() {return name;}
}

File: src/main/scala/mod/iceandshadow3/basics/block/HarvestMethod.scala
package mod.iceandshadow3.basics.block

import javax.annotation.Nullable
import net.minecraftforge.common.ToolType

object HarvestMethod {
	private val lookups = new java.util.HashMap[String, HarvestMethod]
	val HAND = new HarvestMethod("hand", null)
	val PICKAXE = new HarvestMethod("pickaxe")
	val SHOVEL = new HarvestMethod("shovel")
	//Let's not have a spade repeat.
	val AXE = new HarvestMethod("axe")
	val BLADE = new HarvestMethod("sword")
	val SHEAR = new HarvestMethod("shear", null)
	val EXPLOSION = new HarvestMethod("explosion", null)

	def get(name: String) = {
		var retval = lookups.get(name)
		if (retval == null) retval = new HarvestMethod(name, name)
		retval
	}

	def get(tt: ToolType): HarvestMethod = {
		if (tt == null) return HAND
		if (tt eq ToolType.PICKAXE) return PICKAXE
		if (tt eq ToolType.SHOVEL) return SHOVEL
		if (tt eq ToolType.AXE) return AXE
		get(tt.getName)
	}
}

case class HarvestMethod private(name: String, @Nullable classname: String) {
	if (classname != null) HarvestMethod.lookups.putIfAbsent(classname, this)

	def this(string: String) {
		this(string, string)
	}

	override def toString = name
}